File: lib/services/detector.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:camera/camera.dart';
import 'package:flutter/services.dart' show rootBundle;
import 'package:opencv_dart/opencv_dart.dart' as cv;
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';
import 'package:tflite_flutter/tflite_flutter.dart';

import '../models/types.dart';

/// Loads the TFLite model + YuNet face detector, and runs detection on a
/// frame.
///
/// Both face *and* eye localization come from YuNet:
/// - face bbox (cols 0..3 of the [N,15] output)
/// - 5 landmarks (cols 4..13) — we use the right/left eye points to derive
///   eye crops directly. This replaces `haarcascade_eye.xml`, which lost
///   the eye whenever it closed (the cascade was trained on open eyes).
///   YuNet infers landmarks from face geometry, so closed eyes still yield
///   valid eye points.
///
/// Class index map (from notebook):
/// 0=yawn, 1=no_yawn, 2=Closed, 3=Open, 4=front, 5=down
class Detector {
  static const int imgSize = 224;
  static const _yawnIndices = [0, 1]; // yawn vs no_yawn (binary)
  static const _headIndices = [4, 5]; // front vs down (binary)
  static const _eyeIndices = [2, 3];

  // Eye crop side length as a fraction of face width. ~0.30 covers the eye
  // plus enough surrounding skin for the classifier to see lid context.
  static const double _eyeSideFrac = 0.30;

  Interpreter? _interp;
  cv.FaceDetectorYN? _faceDetector;
  (int, int)? _yunetInputSize;

  bool get isReady => _faceDetector != null;
  bool get canClassify => _interp != null;

  /// Asset-bundle init. Only safe on the root isolate (rootBundle requires
  /// the platform-channel binary messenger). Extracts the two model files
  /// to temp and forwards to [initFromPaths].
  Future<void> init({void Function(String)? onProgress}) async {
    onProgress?.call('Extracting models…');
    final tflitePath = await _writeAssetToTemp(
        'assets/models/drowsiness_resnet50v2.tflite',
        'drowsiness_resnet50v2.tflite');
    final yunetPath = await _writeAssetToTemp(
        'assets/models/face_detection_yunet_2023mar.onnx',
        'face_detection_yunet_2023mar.onnx');
    initFromPaths(
      tflitePath: tflitePath,
      yunetPath: yunetPath,
      onProgress: onProgress,
    );
  }

  /// Isolate-safe init. Both files must already exist on disk (the caller
  /// is responsible for extracting them from assets on the root isolate).
  ///
  /// Pass [enableClassification] = false on a detection-only worker — it
  /// avoids loading the ~92 MB TFLite model on isolates that will only
  /// ever call [detectFacesOnly].
  void initFromPaths({
    required String yunetPath,
    String? tflitePath,
    bool enableClassification = true,
    void Function(String)? onProgress,
  }) {
    if (enableClassification) {
      assert(tflitePath != null,
          'tflitePath is required when enableClassification is true');
      onProgress?.call('Loading model…');
      _interp = Interpreter.fromFile(File(tflitePath!));
      _interp!.allocateTensors();
    }

    onProgress?.call('Loading face detector…');
    // Input size is reset per-frame via setInputSize; the (320,320) here is
    // just a starting point.
    _faceDetector = cv.FaceDetectorYN.fromFile(
      yunetPath,
      '',
      (320, 320),
      scoreThreshold: 0.6,
      nmsThreshold: 0.3,
      topK: 50,
    );
    onProgress?.call('Ready');
  }

  Future<String> _writeAssetToTemp(String asset, String name) async {
    final dir = await getTemporaryDirectory();
    final path = p.join(dir.path, name);
    final f = File(path);
    if (!await f.exists()) {
      final bytes = await rootBundle.load(asset);
      await f.writeAsBytes(bytes.buffer.asUint8List());
    }
    return path;
  }

  /// Run detection on a CameraImage frame (mobile path).
  DetectionResult detect(CameraImage image, double confThreshold) {
    if (!isReady) {
      return DetectionResult(
        faces: const [],
        frameWidth: image.width,
        frameHeight: image.height,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    }
    final mat = matFromCameraImage(image);
    if (mat == null) {
      return DetectionResult(
        faces: const [],
        frameWidth: image.width,
        frameHeight: image.height,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    }
    try {
      return detectMat(mat, confThreshold);
    } finally {
      mat.dispose();
    }
  }

  /// YuNet-only detection. Returns boxes + neutral class labels; runs no
  /// classifier passes. Cheap (~10–15 ms on a typical CPU) and intended
  /// to be called every frame so the on-screen bounding box can track the
  /// face fluidly while the slower [detectMat] catches up at a fixed rate.
  DetectionResult detectFacesOnly(cv.Mat mat) {
    if (!isReady) {
      return DetectionResult(
        faces: const [],
        frameWidth: mat.cols,
        frameHeight: mat.rows,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    }
    try {
      final faces = _detectFacesYuNet(mat);
      final out = <FacePrediction>[];
      for (final f in faces) {
        final r = f.rect;
        final fb = FaceBox(r.x, r.y, r.width, r.height);
        final eyeBoxes = <EyePrediction>[];
        for (final pt in [f.rightEye, f.leftEye]) {
          final eb = _eyeBoxFromLandmark(pt, r, mat.cols, mat.rows);
          if (eb != null) {
            // Neutral eye state — overlay will keep showing the last
            // classified state via the merge in drive_page.
            eyeBoxes.add(EyePrediction(eb, EyeClass.open, 0));
          }
        }
        out.add(FacePrediction(
          fb,
          FaceClass.front,
          0,
          eyeBoxes,
          isYawn: false,
          yawnConf: 0,
          isHeadDown: false,
          headPoseConf: 0,
        ));
      }
      return DetectionResult(
        faces: out,
        frameWidth: mat.cols,
        frameHeight: mat.rows,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    } catch (_) {
      return DetectionResult(
        faces: const [],
        frameWidth: mat.cols,
        frameHeight: mat.rows,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    }
  }

  /// Run detection on an already-decoded BGR cv.Mat (desktop path).
  /// Caller owns the Mat lifecycle.
  DetectionResult detectMat(cv.Mat mat, double confThreshold) {
    if (!isReady || !canClassify) {
      return DetectionResult(
        faces: const [],
        frameWidth: mat.cols,
        frameHeight: mat.rows,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    }
    try {
      final faces = _detectFacesYuNet(mat);

      final out = <FacePrediction>[];
      for (final f in faces) {
        final r = f.rect;
        final fb = FaceBox(r.x, r.y, r.width, r.height);
        final probs = _classify(mat, fb);

        // Two independent binaries instead of a single 4-way argmax.
        // The 4-way argmax was hiding yawn signals because front-pose
        // probability dominates the softmax most of the time.
        final yawnP = _renormalized(probs, _yawnIndices);   // [yawn, no_yawn]
        final headP = _renormalized(probs, _headIndices);   // [front, down]
        final isYawn = yawnP[0] > yawnP[1];
        final yawnConf = isYawn ? yawnP[0] : yawnP[1];
        final isHeadDown = headP[1] > headP[0];
        final headConf = isHeadDown ? headP[1] : headP[0];

        // Pick a single display label: surface the most "alarming" signal.
        FaceClass faceClass;
        double faceConf;
        if (isYawn) {
          faceClass = FaceClass.yawn;
          faceConf = yawnConf;
        } else if (isHeadDown) {
          faceClass = FaceClass.down;
          faceConf = headConf;
        } else {
          // both neutral — surface whichever binary is more confident
          faceClass = yawnConf >= headConf ? FaceClass.noYawn : FaceClass.front;
          faceConf = (yawnConf + headConf) / 2;
        }

        // Eye crops come straight from YuNet landmarks — no eye detector
        // pass, so closed eyes still produce valid crops.
        final eyePreds = <EyePrediction>[];
        for (final pt in [f.rightEye, f.leftEye]) {
          final eb = _eyeBoxFromLandmark(pt, r, mat.cols, mat.rows);
          if (eb == null) continue;
          final ep = _classify(mat, eb);
          final ev = _renormalized(ep, _eyeIndices);
          final ei = ev[0] > ev[1] ? 0 : 1;
          eyePreds.add(EyePrediction(
            eb,
            ei == 0 ? EyeClass.closed : EyeClass.open,
            ev[ei],
          ));
        }

        out.add(FacePrediction(
          fb,
          faceClass,
          faceConf,
          eyePreds,
          isYawn: isYawn,
          yawnConf: yawnConf,
          isHeadDown: isHeadDown,
          headPoseConf: headConf,
        ));
      }

      return DetectionResult(
        faces: out,
        frameWidth: mat.cols,
        frameHeight: mat.rows,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    } catch (_) {
      return DetectionResult(
        faces: const [],
        frameWidth: mat.cols,
        frameHeight: mat.rows,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    }
  }

  /// Convert a [CameraImage] to a BGR cv.Mat. Returns null on unsupported
  /// formats. Caller owns the Mat. Exposed so the inference-worker layer
  /// can do the conversion on the root isolate (where the camera frame
  /// lives) before shipping raw bytes to a worker isolate.
  static cv.Mat? matFromCameraImage(CameraImage img) {
    try {
      // BGRA8888 (Windows / iOS when configured): single plane, 4 channels.
      if (img.format.group == ImageFormatGroup.bgra8888) {
        final bytes = img.planes[0].bytes;
        final mat = cv.Mat.fromList(
          img.height,
          img.width,
          cv.MatType.CV_8UC4,
          bytes,
        );
        return cv.cvtColor(mat, cv.COLOR_BGRA2BGR);
      }
      // YUV420 (Android default): convert via opencv_dart helper.
      if (img.format.group == ImageFormatGroup.yuv420) {
        // opencv_dart can take an interleaved YUV NV21 buffer; we build it.
        final y = img.planes[0].bytes;
        final u = img.planes[1].bytes;
        final v = img.planes[2].bytes;
        final nv21 = Uint8List(y.length + u.length + v.length);
        nv21.setAll(0, y);
        for (var i = 0, j = y.length; i < u.length; i++) {
          nv21[j++] = v[i];
          nv21[j++] = u[i];
        }
        final yuvMat = cv.Mat.fromList(
          (img.height * 3 ~/ 2),
          img.width,
          cv.MatType.CV_8UC1,
          nv21,
        );
        return cv.cvtColor(yuvMat, cv.COLOR_YUV2BGR_NV21);
      }
    } catch (_) {}
    return null;
  }

  /// Run model on a 224x224 BGR crop, return raw 6-class softmax outputs.
  ///
  /// IMPORTANT: training (notebook) used `cv2.imread` which returns BGR and
  /// never converted to RGB before fitting. The model therefore expects BGR
  /// inputs — do NOT cvtColor here.
  ///
  /// ResNet50V2 has no internal normalization layer, so we apply
  /// `preprocess_input` manually: scale [0, 255] uint8 to [-1, 1] float.
  Float32List _classify(cv.Mat src, FaceBox box) {
    final roi = src.region(cv.Rect(box.x, box.y, box.w, box.h));
    final resized = cv.resize(roi, (imgSize, imgSize));
    roi.dispose();

    // Build a [1,224,224,3] float32 tensor, applying ResNet50V2's
    // preprocess_input (tf mode: x/127.5 - 1.0).
    final input = Float32List(1 * imgSize * imgSize * 3);
    final raw = resized.data; // Uint8List of HxWx3 in BGR order
    for (var i = 0; i < raw.length; i++) {
      input[i] = raw[i] / 127.5 - 1.0;
    }
    resized.dispose();

    final output = List.filled(6, 0.0).reshape([1, 6]);
    _interp!.run(input.reshape([1, imgSize, imgSize, 3]), output);
    final probs = Float32List(6);
    for (var i = 0; i < 6; i++) {
      probs[i] = (output[0][i] as num).toDouble();
    }
    return probs;
  }

  List<double> _renormalized(Float32List probs, List<int> indices) {
    final subset = indices.map((i) => probs[i]).toList();
    final sum = subset.fold<double>(0, (a, b) => a + b);
    if (sum == 0) return subset.map((_) => 0.0).toList();
    return subset.map((v) => v / sum).toList();
  }

  /// Run YuNet on a BGR Mat and return faces (bbox + eye landmarks) in image
  /// coordinates.
  ///
  /// YuNet's `detect` returns a [N,15] float Mat: 0..3 = x,y,w,h;
  /// 4..5 = right eye, 6..7 = left eye, 8..9 = nose, 10..13 = mouth corners,
  /// 14 = score. We keep bbox + eye points; nose/mouth are unused.
  List<_YuNetFace> _detectFacesYuNet(cv.Mat mat) {
    final size = (mat.cols, mat.rows);
    if (_yunetInputSize == null ||
        _yunetInputSize!.$1 != size.$1 ||
        _yunetInputSize!.$2 != size.$2) {
      _faceDetector!.setInputSize(size);
      _yunetInputSize = size;
    }
    final result = _faceDetector!.detect(mat);
    try {
      final faces = <_YuNetFace>[];
      for (var i = 0; i < result.rows; i++) {
        final x = result.atNum(i, 0).toInt();
        final y = result.atNum(i, 1).toInt();
        final w = result.atNum(i, 2).toInt();
        final h = result.atNum(i, 3).toInt();
        // Clamp to frame bounds — YuNet can return slightly negative
        // coords or boxes that spill past the edge on extreme poses.
        final x0 = x.clamp(0, mat.cols - 1);
        final y0 = y.clamp(0, mat.rows - 1);
        final x1 = (x + w).clamp(0, mat.cols);
        final y1 = (y + h).clamp(0, mat.rows);
        final cw = x1 - x0;
        final ch = y1 - y0;
        if (cw <= 0 || ch <= 0) continue;
        faces.add(_YuNetFace(
          rect: cv.Rect(x0, y0, cw, ch),
          rightEye: (
            result.atNum(i, 4).toDouble(),
            result.atNum(i, 5).toDouble(),
          ),
          leftEye: (
            result.atNum(i, 6).toDouble(),
            result.atNum(i, 7).toDouble(),
          ),
        ));
      }
      return faces;
    } finally {
      result.dispose();
    }
  }

  /// Build a square eye crop around a YuNet eye landmark.
  ///
  /// Side length is a fraction of face width so the crop scales with the
  /// subject's distance from the camera. We clamp the box to the frame *and*
  /// the face rect — if a landmark drifts outside the face on an extreme
  /// pose, this keeps the crop sensible instead of running the classifier
  /// on background.
  FaceBox? _eyeBoxFromLandmark(
    (double, double) pt,
    cv.Rect face,
    int frameW,
    int frameH,
  ) {
    final side = (face.width * _eyeSideFrac).round();
    if (side <= 1) return null;
    final half = side ~/ 2;
    final cx = pt.$1.round();
    final cy = pt.$2.round();
    final faceX1 = face.x + face.width;
    final faceY1 = face.y + face.height;
    int clampI(int v, int lo, int hi) => v < lo ? lo : (v > hi ? hi : v);
    final x0 = clampI(clampI(cx - half, face.x, faceX1 - 1), 0, frameW - 1);
    final y0 = clampI(clampI(cy - half, face.y, faceY1 - 1), 0, frameH - 1);
    final x1 = clampI(clampI(cx + half, face.x, faceX1), 0, frameW);
    final y1 = clampI(clampI(cy + half, face.y, faceY1), 0, frameH);
    final w = x1 - x0;
    final h = y1 - y0;
    if (w <= 1 || h <= 1) return null;
    return FaceBox(x0, y0, w, h);
  }

  void dispose() {
    _interp?.close();
    _faceDetector?.dispose();
  }
}

class _YuNetFace {
  final cv.Rect rect;
  final (double, double) rightEye;
  final (double, double) leftEye;
  _YuNetFace({
    required this.rect,
    required this.rightEye,
    required this.leftEye,
  });
}

File: webapp/src/lib/detector.ts
// Browser port of lib/services/detector.dart.
//
// Pipeline per frame:
//   1. YuNet (onnxruntime-web) → face bbox + right/left eye landmarks
//   2. ResNet50V2 6-class CNN on the face crop → yawn/no_yawn + front/down
//   3. same CNN on each eye crop               → Closed/Open
//
// Class index map: 0=yawn 1=no_yawn 2=Closed 3=Open 4=front 5=down
// Classifier input: [1,224,224,3] NHWC, BGR, ResNet preprocess_input
//   (x/127.5 - 1). The shipped classifier.onnx is INT8-quantized.
// YuNet input:      [1,3,640,640] CHW,  BGR, pixels 0..255.

import * as ort from "onnxruntime-web";
import {
  DetectionResult,
  EyePrediction,
  FaceBox,
  FaceClassName,
  FacePrediction,
  makeResult,
} from "./types";
import { decodeYuNet, YUNET_SIZE } from "./yunet";

const IMG_SIZE = 224;
const EYE_SIDE_FRAC = 0.3;
const YAWN_IDX = [0, 1];
const HEAD_IDX = [4, 5];
const EYE_IDX = [2, 3];

// onnxruntime-web loads its wasm runtime from these files, served same-origin
// from /public/ort. Same-origin is REQUIRED: the app is cross-origin isolated
// (COOP+COEP require-corp, see next.config.mjs) so multi-threaded WASM works,
// and require-corp would block a cross-origin CDN copy of the .wasm.
ort.env.wasm.wasmPaths = "/ort/";

// Use multiple WASM threads. ORT only honours this when the page is
// cross-origin isolated (SharedArrayBuffer available); otherwise it clamps to
// 1 thread on its own, so this is safe everywhere. Cap at 4 to avoid
// oversubscribing phone CPUs. Guarded for SSR (no `navigator` on the server).
if (typeof navigator !== "undefined") {
  ort.env.wasm.numThreads = Math.min(4, navigator.hardwareConcurrency || 2);
}

export class Detector {
  private yunet: ort.InferenceSession | null = null;
  private clf: ort.InferenceSession | null = null;

  // Reusable offscreen canvases for resizing crops.
  private yunetCanvas: HTMLCanvasElement;
  private yunetCtx: CanvasRenderingContext2D;
  private cropCanvas: HTMLCanvasElement;
  private cropCtx: CanvasRenderingContext2D;

  constructor() {
    this.yunetCanvas = document.createElement("canvas");
    this.yunetCanvas.width = YUNET_SIZE;
    this.yunetCanvas.height = YUNET_SIZE;
    this.yunetCtx = this.yunetCanvas.getContext("2d", {
      willReadFrequently: true,
    })!;
    this.cropCanvas = document.createElement("canvas");
    this.cropCanvas.width = IMG_SIZE;
    this.cropCanvas.height = IMG_SIZE;
    this.cropCtx = this.cropCanvas.getContext("2d", {
      willReadFrequently: true,
    })!;
  }

  get isReady() {
    return this.yunet !== null;
  }
  get canClassify() {
    return this.clf !== null;
  }

  async init(onProgress?: (msg: string) => void): Promise<void> {
    const opts: ort.InferenceSession.SessionOptions = {
      executionProviders: ["wasm"],
      graphOptimizationLevel: "all",
    };
    onProgress?.("Loading face detector…");
    this.yunet = await ort.InferenceSession.create(
      "/models/face_detection_yunet_2023mar.onnx",
      opts,
    );
    onProgress?.("Loading model…");
    this.clf = await ort.InferenceSession.create(
      "/models/classifier.onnx",
      opts,
    );
    onProgress?.("Ready");
  }

  /**
   * Run the full pipeline on a video frame.
   * @param source the <video> (or canvas) element to read the current frame from
   * @param srcW   processing width  (frame is sampled at this resolution)
   * @param srcH   processing height
   * @param confThreshold unused at this layer (the alert engine applies it) but
   *                kept for parity; YuNet uses its own fixed score threshold.
   */
  async detect(
    source: CanvasImageSource,
    srcW: number,
    srcH: number,
  ): Promise<DetectionResult> {
    const ts = Date.now();
    if (!this.yunet || !this.clf) {
      return makeResult([], srcW, srcH, ts);
    }

    const yunetFaces = await this.runYuNet(source, srcW, srcH);
    const out: FacePrediction[] = [];

    for (const yf of yunetFaces) {
      // Clamp the face rect to frame bounds.
      const x0 = clampInt(yf.x, 0, srcW - 1);
      const y0 = clampInt(yf.y, 0, srcH - 1);
      const x1 = clampInt(yf.x + yf.w, 0, srcW);
      const y1 = clampInt(yf.y + yf.h, 0, srcH);
      const fw = x1 - x0;
      const fh = y1 - y0;
      if (fw <= 0 || fh <= 0) continue;
      const faceBox: FaceBox = { x: x0, y: y0, w: fw, h: fh };

      const probs = await this.classify(source, faceBox, srcW, srcH);

      const yawnP = renorm(probs, YAWN_IDX);
      const headP = renorm(probs, HEAD_IDX);
      const isYawn = yawnP[0] > yawnP[1];
      const yawnConf = isYawn ? yawnP[0] : yawnP[1];
      const isHeadDown = headP[1] > headP[0];
      const headConf = isHeadDown ? headP[1] : headP[0];

      let faceClass: FaceClassName;
      let faceConf: number;
      if (isYawn) {
        faceClass = "yawn";
        faceConf = yawnConf;
      } else if (isHeadDown) {
        faceClass = "down";
        faceConf = headConf;
      } else {
        faceClass = yawnConf >= headConf ? "no_yawn" : "front";
        faceConf = (yawnConf + headConf) / 2;
      }

      const eyes: EyePrediction[] = [];
      for (const pt of [yf.rightEye, yf.leftEye]) {
        const eb = eyeBoxFromLandmark(pt, faceBox, srcW, srcH);
        if (!eb) continue;
        const ep = await this.classify(source, eb, srcW, srcH);
        const ev = renorm(ep, EYE_IDX);
        const closed = ev[0] > ev[1];
        eyes.push({
          box: eb,
          eyeClass: closed ? "Closed" : "Open",
          conf: closed ? ev[0] : ev[1],
        });
      }

      out.push({
        box: faceBox,
        faceClass,
        conf: faceConf,
        isYawn,
        yawnConf,
        isHeadDown,
        headPoseConf: headConf,
        eyes,
      });
    }

    return makeResult(out, srcW, srcH, ts);
  }

  private async runYuNet(
    source: CanvasImageSource,
    srcW: number,
    srcH: number,
  ) {
    // Stretch-resize the frame into the 640x640 input (matches OpenCV
    // FaceDetectorYN.setInputSize, which also just resizes, no letterbox).
    this.yunetCtx.drawImage(source, 0, 0, YUNET_SIZE, YUNET_SIZE);
    const data = this.yunetCtx.getImageData(0, 0, YUNET_SIZE, YUNET_SIZE).data;

    // CHW, BGR, 0..255.
    const plane = YUNET_SIZE * YUNET_SIZE;
    const input = new Float32Array(3 * plane);
    for (let i = 0; i < plane; i++) {
      const r = data[i * 4 + 0];
      const g = data[i * 4 + 1];
      const b = data[i * 4 + 2];
      input[i] = b; // B plane
      input[plane + i] = g; // G plane
      input[2 * plane + i] = r; // R plane
    }
    const tensor = new ort.Tensor("float32", input, [1, 3, YUNET_SIZE, YUNET_SIZE]);
    const result = await this.yunet!.run({
      [this.yunet!.inputNames[0]]: tensor,
    });

    const outs: Record<string, Float32Array> = {};
    for (const name of this.yunet!.outputNames) {
      outs[name] = result[name].data as Float32Array;
    }
    const faces = decodeYuNet(outs);

    // Scale 640x640 coords back to the source frame (per-axis).
    const sx = srcW / YUNET_SIZE;
    const sy = srcH / YUNET_SIZE;
    return faces.map((f) => ({
      x: f.x * sx,
      y: f.y * sy,
      w: f.w * sx,
      h: f.h * sy,
      score: f.score,
      rightEye: [f.rightEye[0] * sx, f.rightEye[1] * sy] as [number, number],
      leftEye: [f.leftEye[0] * sx, f.leftEye[1] * sy] as [number, number],
    }));
  }

  /** Run the 6-class CNN on a crop, return raw softmax probabilities. */
  private async classify(
    source: CanvasImageSource,
    box: FaceBox,
    _srcW: number,
    _srcH: number,
  ): Promise<Float32Array> {
    // Draw the crop region resized to 145x145.
    this.cropCtx.drawImage(
      source,
      box.x,
      box.y,
      box.w,
      box.h,
      0,
      0,
      IMG_SIZE,
      IMG_SIZE,
    );
    const data = this.cropCtx.getImageData(0, 0, IMG_SIZE, IMG_SIZE).data;

    // NHWC, BGR, ResNet50V2 preprocess_input: x/127.5 - 1 → [-1, 1].
    const input = new Float32Array(IMG_SIZE * IMG_SIZE * 3);
    for (let i = 0; i < IMG_SIZE * IMG_SIZE; i++) {
      const r = data[i * 4 + 0];
      const g = data[i * 4 + 1];
      const b = data[i * 4 + 2];
      input[i * 3 + 0] = b / 127.5 - 1;
      input[i * 3 + 1] = g / 127.5 - 1;
      input[i * 3 + 2] = r / 127.5 - 1;
    }
    const tensor = new ort.Tensor("float32", input, [1, IMG_SIZE, IMG_SIZE, 3]);
    const result = await this.clf!.run({
      [this.clf!.inputNames[0]]: tensor,
    });
    return result[this.clf!.outputNames[0]].data as Float32Array;
  }

  dispose() {
    this.yunet?.release();
    this.clf?.release();
    this.yunet = null;
    this.clf = null;
  }
}

function renorm(probs: Float32Array, indices: number[]): number[] {
  const subset = indices.map((i) => probs[i]);
  const sum = subset.reduce((a, b) => a + b, 0);
  if (sum === 0) return subset.map(() => 0);
  return subset.map((v) => v / sum);
}

function clampInt(v: number, lo: number, hi: number): number {
  v = Math.round(v);
  return v < lo ? lo : v > hi ? hi : v;
}

function eyeBoxFromLandmark(
  pt: [number, number],
  face: FaceBox,
  frameW: number,
  frameH: number,
): FaceBox | null {
  const side = Math.round(face.w * EYE_SIDE_FRAC);
  if (side <= 1) return null;
  const half = Math.floor(side / 2);
  const cx = Math.round(pt[0]);
  const cy = Math.round(pt[1]);
  const faceX1 = face.x + face.w;
  const faceY1 = face.y + face.h;
  const c = (v: number, lo: number, hi: number) =>
    v < lo ? lo : v > hi ? hi : v;
  const x0 = c(c(cx - half, face.x, faceX1 - 1), 0, frameW - 1);
  const y0 = c(c(cy - half, face.y, faceY1 - 1), 0, frameH - 1);
  const x1 = c(c(cx + half, face.x, faceX1), 0, frameW);
  const y1 = c(c(cy + half, face.y, faceY1), 0, frameH);
  const w = x1 - x0;
  const h = y1 - y0;
  if (w <= 1 || h <= 1) return null;
  return { x: x0, y: y0, w, h };
}
